[c10d] Fix CudaEventCache for dangling references

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144496

Reported in https://github.com/pytorch/pytorch/issues/143470, we have a dangling references in `CudaEventCache`. So we want to fix it.
1. We add a unit test to repro the issue mentioned in the issue.
2. Instead of converting variables to shared pointers as suggested in the issue, we then make the cache itself a shared pointer. So if the thread creates the cache dies before all events get recycled, the cache is still there until the last CudaEvent get deleted. (thanks for the suggestion from @kwen2501 )

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
Fix CudaEventCache for dangling references

[ghstack-poisoned]

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -438,6 +438,32 @@ def test_restart_pg(self):
 
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    def test_cuda_event_cache(self):
```
can you make the test name more descriptive or at least add a comment for what the test is testing?

iiuc the point is to test some particular race on thread destruction

fduwjj replied:
yes. Let me update it.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -847,9 +849,10 @@ ProcessGroupNCCL::CUDAEventCache& ProcessGroupNCCL::CUDAEventCache::get(
         std::forward_as_tuple(device),
-        std::forward_as_tuple());
+        std::forward_as_tuple(
+            std::make_shared<ProcessGroupNCCL::CUDAEventCache>()));
   }
-  return it->second;
+  return *(it->second);
```
This function may need to return a `std::shared_ptr<ProcessGroupNCCL::CUDAEventCache>>` than a reference. That is, if we start to use `std::shared_ptr` we may probably need to use it all the way. (That is, the original `return it->second` can remain the same.)

fduwjj replied:
Is this really needed? I mean after calling `get`, the callsite will do dereference anyway.

kwen2501 replied:
I can't comment on whether it is really needed, but seems a general practice to not mix the use of shared_ptr with reference.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -806,11 +806,13 @@ ProcessGroupNCCL::CUDAEventCache::CUDAEventCache() = default;
   // register the deleter as a callback when the WorkNCCL object is destroyed.
-  auto deleter = [this, timing](at::cuda::CUDAEvent* event) {
-    std::lock_guard<std::mutex> lock(this->cacheMutex_);
+  auto deleter = [cache =
+                      std::shared_ptr<ProcessGroupNCCL::CUDAEventCache>(this),
```
This may create a separate reference counting than the original one created (by `std::make_shared`). That may be the reason we see double free. We may need to use [`share_from_this`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this/shared_from_this) instead, to make it one single count.

kwen2501 replied:
Just writing down my thought process:
each deleter will keep a reference to the `CUDAEventCache`. And the deleter goes with the life of the `CUDAEvent`. That is, there would be (roughly) as many references to the cache as there are `CUDAEvent`s. When the `CUDAEvent` is destructed, the deleter is gone, releasing one ref. After all `CUDAEvent`s being destructed, the ref count drops to 0, then the cache would be destructed -- and so we keep the life of the cache longer than the thread that creates it. The map is still gone as the creating thread dies, but since the map holds only a shared_ptr to the cache, it would not destruct the cache, but only reducing the ref count by 1.

fduwjj replied:
Yeah that was what I planned to do originally. I misunderstood how the ref count works here. (I thought using `std::shared_ptr` will reuse the original ref count) Using `[share_from_this](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this/shared_from_this)` indeed works. And now the unit test passed.

fduwjj committed:
Update on "Fix CudaEventCache for dangling references"




cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Fix CudaEventCache for dangling references"


Reported in https://github.com/pytorch/pytorch/issues/143470, we have a dangling references in `CudaEventCache`. So we want to fix it.
1. We add a unit test to repro the issue mentioned in the issue.
2. Instead of converting variables to shared pointers as suggested in the issue, we then make the cache itself a shared pointer. So if the thread creates the cache dies before all events get recycled, the cache is still there until the last CudaEvent get deleted. (thanks for the suggestion from kwen2501 )

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Fix CudaEventCache for dangling references"


Reported in https://github.com/pytorch/pytorch/issues/143470, we have a dangling references in `CudaEventCache`. So we want to fix it.
1. We add a unit test to repro the issue mentioned in the issue.
2. Instead of converting variables to shared pointers as suggested in the issue, we then make the cache itself a shared pointer. So if the thread creates the cache dies before all events get recycled, the cache is still there until the last CudaEvent get deleted. (thanks for the suggestion from kwen2501 )

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -847,7 +848,8 @@ ProcessGroupNCCL::CUDAEventCache& ProcessGroupNCCL::CUDAEventCache::get(
         it,
         std::piecewise_construct,
         std::forward_as_tuple(device),
-        std::forward_as_tuple());
+        std::forward_as_tuple(
+            std::make_shared<ProcessGroupNCCL::CUDAEventCache>()));
```
Just a note: we may be able to simplify it in future now that we are not emplacing a mutex into the map. (Not required for this PR)

fduwjj replied:
oh right, we can use just emplace now. let me try.

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -438,6 +438,35 @@ def test_restart_pg(self):
+            dist.all_reduce(t)
+            dist.all_reduce(t)
+
+        os.environ["TORCH_NCCL_CUDA_EVENT_CACHE"] = "1"
```
nit: caution when setting env in unit tests -- it can carry over to other unit tests if the process is not restarted -- like we did for `test_c10d_ops_nccl.py` to reuse PGs. `test_c10d_nccl` is fine, but can be refactored in future.

fduwjj replied:
I can add a ENV reset, lol.

fduwjj replied:
Hmm since we launch it anyway so I think this should be fine now. And we want to remove the ENV maybe before next PT release.

kwen2501 reviewed (approved):
Looks nice and clean! Thanks for implementing the idea! 
nit: would be nice if we could put down some comments (like [this](https://github.com/pytorch/pytorch/pull/144496#discussion_r1910144873)) in the code to validate the thought process so that future developers are aware.

fduwjj committed:
Update on "[c10d] Fix CudaEventCache for dangling references"


Reported in https://github.com/pytorch/pytorch/issues/143470, we have a dangling references in `CudaEventCache`. So we want to fix it.
1. We add a unit test to repro the issue mentioned in the issue.
2. Instead of converting variables to shared pointers as suggested in the issue, we then make the cache itself a shared pointer. So if the thread creates the cache dies before all events get recycled, the cache is still there until the last CudaEvent get deleted. (thanks for the suggestion from kwen2501 )

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
